pkg/pillar/zedUpload/datastore_req.go

```diff
@@ -196,6 +198,29 @@ func (req *DronaRequest) updateOsize(size int64) {
 	req.objectSize = size
 }
 
+// Update the object size and chunk data
+func (req *DronaRequest) updateChunkData(chunkData ChunkData) {
+	req.Lock()
+	defer req.Unlock()
+	req.chunkInfo = chunkData
+	req.result <- req
+}
+
+// update chunk details on chunk
+func (req *DronaRequest) postChunk(chunkData ChunkData) {
+	req.Lock()
+	defer req.Unlock()
+	req.chunkInfo = chunkData
+	req.result <- req
```

Comment:
Did you have this pattern earlier?

What prevents req.chunkInfo from being overwritten when the channel is busy or before the reader is accessing req (the channel takes a pointer to the DronaRequest not its content).
Basically one go-routine calls postChunk(X), thus req.ChunkInfo is set to X, req posted to channel, return. Then that caller proceeds to call postChunk(Y), thus req.ChunkInfo is set to Y. This can happen before/while the reader of the channel reads req and accesses chunkInfo.

Ah - didn't see the Lock/Unlock. But how do you avoid deadlock with the locks?

Is there existing code which uses req.result this way? grepping in eve and zedcloud I can't find any.

Reply:
Actually even with the locks it can get clobbered in the example. By the time the DronaRequest pointer is read from the channel but before that go-routine has called GetChunkDetauls the writer could jave called postChunk(Y) and updated req.ChunkInfo. Thus X would be lost.

You need to send the actual data on a channel to avoid loosing it, or you need to build a queue of the data and not just a single ChunkInfo. If req.request isn't used anywhere I suggest we delete it, because it seems like a hazard which is easy to trip on.
And if the request

Reply:
as you suggested, can i use a chunkInfo as a map of <int, ChunkData>, where it has the key which denotes the order of accessing the chunks(1,2,3..).
the moment the map is added with new chunk, the req.result is notified, so that caller will fetch from the map with the method GetChunkDetails by specifying the key(1,2,3,4...), as when the chunks are read successfully at the caller end, will remove the entry from the chunkInfo.

 let me know your inputs, so that I can start work on it and wrap this up.

Reply:
1. Maps aren't natural for this since you'd need to search for the next key. (Plus maps are not safe to be used by different goroutines. But you can use the base package (LockedStringMap type) if you really needed a map.)
2. A slice would work as long as you add locks and make sure the elements in the slice are garbage collected (the slice has indicies/len referring an array, and elements can stay around even if the len is zero.). Plus needs locking.
3. The natural thing in go for a fixed length queue is a buffered channel. Safe to use in different go-routines.

So I'd suggest replacing the unused result channel of pointers with a channel of type ChunkData (called chunkInfoChan). FWIW I think you can make it have a len 1 to be more efficient; the reader and writer don't have to rendez-vous on the channel but you can queue one ChunkData on it.

With that you end up with just 
func (req *DronaRequest) postChunk(chunkData ChunkData) {
	req.chunkInfoChan <- chunkData
	req.result <- req
}

func (req *DronaRequest) GetChunkDetails() (int64, []byte, bool) {
	chunkInfo <- req.chunkInfoChan
	return chunkInfo.Size, chunkInfo.Chunk, chunkInfo.EOF
}

BTW: I think the updateChunkData function is no longer used.